### PR TITLE
Fix documentation panel display logic of the completer

### PR DESCRIPTION
**Pull request by trungleduc**
## References
Closes https://github.com/jupyterlab/jupyterlab/issues/15018

## Code changes

- `CompleterModel.resolveItem` now accepts both the completion item and its index but resolving by index will be deprecated in the future.
- Fix the Regex issue in the default `itemWidthHeuristic`

## User-facing changes

N/A

## Backwards-incompatible changes

The interface `Completer.IModel` has been changed but still is compatible with previous versions.

### CODE REVIEW

File: packages/completer/src/model.ts [hunk 1]
```diff
@@ -459,6 +472,28 @@ export class CompleterModel implements Completer.IModel {
...
+   * @return See `resolveItem` method
+   */
+  private _resolveItemByIndex(
+    activeIndex: number
+  ): Promise<CompletionHandler.ICompletionItem | null> | undefined {
+    if (!this.completionItems) {
+      return undefined;
+    }
+
+    let completionItems = this.completionItems();
```
[the remark this answers is not in the document]
Comment (trungleduc) [reply packages/completer/src/model.ts:489]: By doing so, we need to change the `Completer.IModel` interface. Do you think we should add a new getter `completionItemsWithIndex` for this?

[the remark this answers is not in the document]
Comment (krassowski) [reply packages/completer/src/model.ts:489]: No, I would not go the route of adding a new getter because the state already is hard to follow. An alternative would be a private Map or WeakMap from processed to original item.

File: packages/completer/src/widget.ts [hunk 2]
```diff
@@ -1219,9 +1223,11 @@ export namespace Completer {
      * Get a heuristic for the width of an item.
      */
     itemWidthHeuristic(item: CompletionHandler.ICompletionItem): number {
-      return (
-        item.label.replace(/<\?mark>/g, '').length + (item.type?.length || 0)
-      );
+      // Get the label text for both escaped and unescaped text.
+      const labelText = item.label
+        .replace(/<(\/)?mark>/g, '')
+        .replace(/&lt;(\/)?mark&gt;/g, '');
```
[the remark this answers is not in the document]
Comment (trungleduc) [reply packages/completer/src/widget.ts]: Indeed it's because of using the processed items in the resolution. Thanks!

File: packages/completer/src/model.ts [hunk 3]
```diff
@@ -365,65 +373,106 @@ export class CompleterModel implements Completer.IModel {
...
   }
 
   /**
-   * Lazy load missing data of item at `activeIndex`.
-   * @param {number} activeIndex - index of item
+   * Lazy load missing data of an item.
+   * @param indexOrValue - the item or its index
+   * @remarks
+   * Resolving item by index will be deprecated in
+   * the next major release.
```
Comment (krassowski) [inline packages/completer/src/model.ts:429]: Should we be more specific here? Just a suggestion on wording, if you meant other versions that is fine with me too.

```suggestion
   * the JupyterLab 5.0 and removed in JupyterLab 6.0.
```

**Review by krassowski — APPROVED** [review]
Thank you @trungleduc, I pushed one commit with a test of `itemWidthHeuristic` (thanks for the fix!) and added a minor suggestion on the deprecation notice.

**Comment by trungleduc** [comment]
Thanks @krassowski
